packages/cli/src/google/download.ts

```diff
@@ -19,180 +18,169 @@ const writeDownload = async (url: string, dest: fs.PathLike) => {
-			hasUnicodeSubsets = true;
+type Variant = StaticVariant | VariableVariant;
+
+const isAbsoluteURL = (url: string): boolean => {
+	return /^https?:\/\//.test(url);
+};
+
+const getStaticVariantList = (variants: FontVariants): StaticVariant[] => {
+	const variantList: StaticVariant[] = [];
+	for (const [weight, styles] of Object.entries(variants)) {
+		for (const [style, subsets] of Object.entries(styles)) {
+			for (const [subset, { url }] of Object.entries(subsets)) {
+				const props = {
+					kind: 'static' as const,
+					weight: Number(weight),
+					style,
+					subset: subset.replace('[', '').replace(']', ''),
+				};
+				if (isAbsoluteURL(url.woff2)) {
+					variantList.push({
+						...props,
+						url: url.woff2,
+						extension: 'woff2',
+					});
+				}
+				if (isAbsoluteURL(url.woff)) {
+					variantList.push({
+						...props,
+						url: url.woff,
+						extension: 'woff',
+					});
+				}
+			}
 		}
 	}
+	return variantList;
+};
 
-	// If true, we need to download the woff2 files from V1. Else remove all woff2 files
-	if (!hasUnicodeSubsets) {
-		downloadURLPairsV1 = downloadURLPairsV1.filter(
-			(pair) => pair[0][4] !== 'woff2'
-		);
+const getVariableVariantList = (
+	variants: FontVariantsVariable
+): VariableVariant[] => {
+	const variantList: VariableVariant[] = [];
+	for (const [axes, styles] of Object.entries(variants)) {
+		for (const [style, subsets] of Object.entries(styles)) {
+			for (const [subset, url] of Object.entries(subsets)) {
+				if (isAbsoluteURL(url)) {
+					variantList.push({
+						kind: 'variable',
+						axes,
+						style,
+						subset: subset.replace('[', '').replace(']', ''),
+						url: url,
+						extension: 'woff2',
+					});
+				}
+			}
+		}
 	}
+	return variantList;
+};
 
-	// V1 { url, dest } pairs
-	const linksV1 = downloadURLPairsV1.map((pair) => {
-		const types = pair[0];
-		const dest = makeFontDownloadPath(
+interface DownloadLinks {
+	url: string;
+	dest: string;
+}
+
+const getDownloadPath = (
+	id: string,
+	variant: Variant,
+	opts: BuildOptions
+): string => {
+	if (variant.kind === 'static') {
+		return makeFontDownloadPath(
 			opts.dir,
 			id,
-			types[2],
-			Number(types[0]),
-			types[1],
-			types[4]
+			variant.subset,
+			variant.weight,
+			variant.style,
+			variant.extension
 		);
-		const url = pair[1];
-		return {
-			url,
-			dest,
-		};
-	});
-
-	// V2 { url, dest } pairs
-	const linksV2 = downloadURLPairsV2.map((pair) => {
-		const types = pair[0];
-		const dest = makeFontDownloadPath(
+	}
+	if (variant.kind === 'variable') {
+		return makeVariableFontDownloadPath(
 			opts.dir,
 			id,
-			types[2].replace('[', '').replace(']', ''),
-			Number(types[0]),
-			types[1],
-			types[4]
+			variant.subset,
+			variant.axes,
+			variant.style
 		);
-
-		const url = pair[1];
-		return { url, dest };
-	});
-
-	const links = [...linksV1, ...linksV2];
-	return links;
+	}
+	// Makes sure that all variants are handled above.
+	return variant;
```

Comment:
This should never happen, right? Maybe a throw is better here?

Reply:
Yes, this should never happen, and this return **compile-time ensures** that. Try to remove one of the if's, and you will get a type error! This ensures that if for whatever reason a new type is added to the union, it will be handled here (... except if it's a string 😬). It works because 'variant' gets narrowed to the 'never' type, which can be assigned to anything (as it will never run). If a new kind were to be added (that's not compatible with string), it will be a type error.

I agree that this is not very intuitive though, so if you do not like it, I will change it to a throw :)

(or we could use something like this https://github.com/aikoven/assert-never)

Reply:
Ooh, I see. Learnt a new pattern here.

I'm a little conflicted since whilst I like the type-safety, I think it hurts immediate readability. A throw conveys the "never should run" nature better imo, but I'm not against you wrapping it in a helper like `assert-never` to get the best of both worlds (although I don't think you should install it as a dep and write your own).

Reply:
Added an `assertNever` utility function, how is it now?

Reply:
https://github.com/aikoven/assert-never/blob/774509a47569ce42e99d3ae580ca11736e73a297/index.ts#L31 - as a sanity check, we should throw in the assertion. 👍 Should be perfect then.

Reply:
Done!

(I would say that throw can't ever execute, but... unfortunately TypeScript's type system is not entirely immune to such things (`any`), so it's valid concern.